opentelemetry-api/src/opentelemetry/trace/sampling.py

```diff
@@ -0,0 +1,128 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+from typing import Dict, Mapping, Optional, Sequence
+
+# pylint: disable=unused-import
+from opentelemetry.trace import Link, SpanContext
+from opentelemetry.util.types import AttributeValue
+
+
+class Decision:
+    """A sampling decision as applied to a newly-created Span.
+
+    Args:
+        sampled: Whether the `Span` should be sampled.
+        attributes: Attributes to add to the `Span`.
+    """
+
+    def __repr__(self) -> str:
+        return "{}({}, attributes={})".format(
+            type(self).__name__, str(self.sampled), str(self.attributes)
+        )
+
+    def __init__(
+        self,
+        sampled: bool = False,
+        attributes: Mapping[str, "AttributeValue"] = None,
+    ) -> None:
+        self.sampled: bool
+        self.attributes: Dict[str, "AttributeValue"]
+
+        self.sampled = sampled
+        if attributes is None:
+            self.attributes = {}
+        else:
+            self.attributes = dict(attributes)
+
+
+class Sampler(abc.ABC):
+    @abc.abstractmethod
+    def should_sample(
+        self,
+        parent_context: Optional["SpanContext"],
+        trace_id: int,
+        span_id: int,
+        name: str,
+        links: Optional[Sequence["Link"]] = None,
+    ) -> "Decision":
+        pass
+
+
+class StaticSampler(Sampler):
+    """Sampler that always returns the same decision."""
+
+    def __init__(self, decision: "Decision"):
+        self._decision = decision
+
+    def should_sample(
+        self,
+        parent_context: Optional["SpanContext"],
+        trace_id: int,
+        span_id: int,
+        name: str,
+        links: Optional[Sequence["Link"]] = None,
+    ) -> "Decision":
+        return self._decision
+
+
+class ProbabilitySampler(Sampler):
+    def __init__(self, rate: float):
+        self._rate = rate
+        self._bound = self.get_bound_for_rate(self._rate)
+
+    # The sampler checks the last 8 bytes of the trace ID to decide whether to
+    # sample a given trace.
+    CHECK_BYTES = 0xFFFFFFFFFFFFFFFF
+
+    @classmethod
+    def get_bound_for_rate(cls, rate: float) -> int:
+        return round(rate * (cls.CHECK_BYTES + 1))
+
+    @property
+    def rate(self) -> float:
+        return self._rate
+
+    @rate.setter
+    def rate(self, new_rate: float) -> None:
+        self._rate = new_rate
+        self._bound = self.get_bound_for_rate(self._rate)
+
+    @property
+    def bound(self) -> int:
+        return self._bound
+
+    def should_sample(
+        self,
+        parent_context: Optional["SpanContext"],
+        trace_id: int,
+        span_id: int,
+        name: str,
+        links: Optional[Sequence["Link"]] = None,
+    ) -> "Decision":
+        if parent_context is not None:
+            return Decision(parent_context.trace_options.recorded)
+
+        return Decision(trace_id & self.CHECK_BYTES < self.bound, {})
```

Comment:
If we keep the current code of Decision's `__init__`, we should use `None` rather than `{}` here to avoid an additional dict copy.

Reply:
Changed in b653f3e, but I think more important than the dict copy is whether the sampler should inherit any "default" attributes that the base class might attach. I think this sampler should, but it's a moot point since we don't have any default attributes yet.